### PR TITLE
Refactor function CRD

### PR DESCRIPTION
**Pull request by m00g3n**
**Description**

Changes proposed in this pull request:

- the version of Function CRD was bumped to `v1alpha2`
- the Repository CRD is no longer supported

**Related issue(s)**

See also: https://github.com/kyma-project/kyma/issues/13828

### CODE REVIEW

File: components/function-controller/internal/controllers/serverless/build_resources.go [hunk 1]
```diff
@@ -131,3 +133,13 @@ func (r *FunctionReconciler) mergeLabels(labelsCollection ...map[string]string)
 	}
 	return result
 }
+
+func (r *FunctionReconciler) getPackageConfigVolumeMountsForRuntime(rtm serverlessv1alpha2.Runtime) []corev1.VolumeMount {
```
Comment (pPrecel) [inline components/function-controller/internal/controllers/serverless/build_resources.go:137]: is this function used anywhere? 🤔

File: components/function-controller/internal/controllers/serverless/fixtures_test.go [hunk 2]
```diff
@@ -19,20 +19,25 @@ func newTestGitFunction(namespace, name string, minReplicas, maxReplicas int, co
...
-			Type:    serverlessv1alpha1.SourceTypeGit,
-			Source:  fmt.Sprintf("%s-%d", name, suffix),
-			Runtime: serverlessv1alpha1.Nodejs12,
-			Repository: serverlessv1alpha1.Repository{
-				BaseDir:   "/",
-				Reference: "main",
+		Spec: serverlessv1alpha2.FunctionSpec{
+			Source: serverlessv1alpha2.Source{
+				GitRepository: serverlessv1alpha2.GitRepositorySource{
+					//URL: fmt.Sprintf("%s-%d", name, suffix),
```
Comment (pPrecel) [inline components/function-controller/internal/controllers/serverless/fixtures_test.go:31]: can we remove it? or is this something that will be changed before merging?

File: components/function-controller/pkg/apis/serverless/v1alpha2/function_validation_test.go [hunk 3]
```diff
@@ -0,0 +1,504 @@
...
+	"github.com/vrischmann/envconfig"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+
+	"github.com/onsi/gomega"
+)
+
+func TestFunctionSpec_validateResources(t *testing.T) {
```
Comment (pPrecel) [inline components/function-controller/pkg/apis/serverless/v1alpha2/function_validation_test.go:16]: Good Job! 🥇

File: components/function-controller/pkg/apis/serverless/v1alpha2/function_defaults_test.go [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+
+	"github.com/vrischmann/envconfig"
+
+	"github.com/onsi/gomega"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+func TestSetDefaults(t *testing.T) {
```
Comment (pPrecel) [inline components/function-controller/pkg/apis/serverless/v1alpha2/function_defaults_test.go:18]: Well done! 👍

File: components/function-controller/internal/controllers/serverless/gitops.go [hunk 5]
```diff
@@ -3,11 +3,21 @@ package serverless
...
 	"fmt"
 
-	ctrl "sigs.k8s.io/controller-runtime"
-
 	"github.com/kyma-project/kyma/components/function-controller/internal/git"
+	serverlessv1alpha2 "github.com/kyma-project/kyma/components/function-controller/pkg/apis/serverless/v1alpha2"
+	ctrl "sigs.k8s.io/controller-runtime"
 )
 
+//TODO refactor - move to system state
```
Comment (pPrecel) [inline components/function-controller/internal/controllers/serverless/gitops.go:11]: Do you plan to work on it in this story or the next one?

File: components/function-controller/internal/controllers/serverless/gitops.go [hunk 6]
```diff
@@ -3,11 +3,21 @@ package serverless
...
+func (r *FunctionReconciler) syncRevision(instance *serverlessv1alpha2.Function, options git.Options) (string, error) {
```
Comment (pPrecel) [inline components/function-controller/internal/controllers/serverless/gitops.go:12]: Is this func used anywhere?

**Review by pPrecel — COMMENTED** [review]
I found a few places to improve. But nothing critical. Looks great even without those fixes ;)  Would be great if you share with me what you think about this!

Now I need some time to manually test this functionality on a cluster and then I can give you an approval!

File: components/function-controller/config/crd/bases/serverless.kyma-project.io_functions.yaml [hunk 7]
```diff
@@ -4,7 +4,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.2.9
```
[the remark this answers is not in the document]
Comment (m00g3n) [reply components/function-controller/config/crd/bases/serverless.kyma-project.io_functions.yaml:7]: Unintentional - my version of `controller-gen` was outdated.

File: components/function-controller/config/crd/bases/serverless.kyma-project.io_gitrepositories.yaml [hunk 8]
```diff
@@ -4,7 +4,7 @@ apiVersion: apiextensions.k8s.io/v1
 kind: CustomResourceDefinition
 metadata:
   annotations:
-    controller-gen.kubebuilder.io/version: v0.2.9
+    controller-gen.kubebuilder.io/version: v0.2.4
   creationTimestamp: null
   name: gitrepositories.serverless.kyma-project.io
```
[the remark this answers is not in the document]
Comment (m00g3n) [reply components/function-controller/config/crd/bases/serverless.kyma-project.io_gitrepositories.yaml:9]: same as https://github.com/kyma-project/kyma/pull/14234#discussion_r894284374

File: components/function-controller/internal/controllers/serverless/gitops.go (hunk 5 above)
Comment (m00g3n) [reply components/function-controller/internal/controllers/serverless/gitops.go:11]: This is a leftover, will delete it.

File: components/function-controller/internal/controllers/serverless/gitops.go (hunk 6 above)
Comment (m00g3n) [reply components/function-controller/internal/controllers/serverless/gitops.go:12]: Not used, a leftover, will delete, many thx

**Review by dbadura — COMMENTED** [review]
Can we add short name `fn`. It would be useful when working with `kubectl`, you could type: `kubectl get fn`?

File: components/function-controller/config/crd/bases/serverless.kyma-project.io_functions.yaml [hunk 9]
```diff
@@ -26,7 +26,7 @@ spec:
     - jsonPath: .status.conditions[?(@.type=='Running')].status
```
Comment (Cortey) [inline components/function-controller/config/crd/bases/serverless.kyma-project.io_functions.yaml:26]: Include config samples and think about extending rest of the resources.

**Comment by pPrecel** [comment]
This PR will be extended by the functionality described here:
https://github.com/m00g3n/kyma/pull/3

**Comment by pbochynski** [comment]
@m00g3n  are you still working on it?

**Pull request closed without merging**